source/kraken/fill_disruption_from_chaos.cpp

```diff
@@ -402,25 +453,248 @@ struct functor_add_vj {
+
+              //check the date if adapted_vj is inside impact periode
+              if (is_impacted(*adapted_vj)) {
+                  // we check if there are any of journey pattern points of this adapted vj that belongs to
+                  // impacted stop points list
+                  if (is_updated_with_impacted_stop_points(adapted_vj)) {
+                      // if this adpated_vj has no JPP in this stop point list, it means that its theoretical
+                      // vj is also impacted by this impact.
+                      push_back_unique(impact->impacted_journey_patterns, adapted_vj->journey_pattern);
+
+                      auto has_no_impact = std::none_of(std::begin(adapted_vj->impacted_by),
+                              std::end(adapted_vj->impacted_by),
+                              [&](boost::weak_ptr<navitia::type::new_disruption::Impact>& weak_ptr ){
+                          if(auto ptr = weak_ptr.lock()) {
+                              return ptr.get() == impact.get();
+                          }
+                          return false;
+                      });
+                      if (has_no_impact) {
+                          adapted_vj->impacted_by.push_back(impact);
+                      }
+                  }
+                  // we find an impacted vj for the vj, we don't need to bother with it, it will be update later
+                  return true;
+              };
+          }
+
+          // If adapted vj doesn't exist, we get_or_create a new jp with a new list of stop points
+          // after the application of impact
+          auto new_jp = get_or_create_journey_pattern(*jp);
+
+          // It's not possible that jp and new_jp are same, beacuse they cannot have the same stop points
+          assert(new_jp != jp);
+
+          functor_copy_vj_to_jp{impact, new_jp, impacted_stop_points, pt_data, meta}(*vj);
+
+          // The new journey_pattern is linked to the impact
+          push_back_unique(impact->impacted_journey_patterns, new_jp);
+      }
+      return true;
+  }
+};
+
+// Impacts can only be collected vj,
+// this is a convenient functor to collect all impacts that are applied on a jp
+// You can call jp->for_each_vehicle_journey(ImpactsCollector{}) to get all impacts
+struct ImpactsCollector {
+public:
+    template<typename T>
+    bool operator()(std::unique_ptr<T>& vj) {
+        boost::for_each(vj->impacted_by, [&](const typename decltype(vj->impacted_by)::value_type& weak_ptr){
+                if(auto shared_ptr = weak_ptr.lock()) {
+                    jp_impacts.insert(shared_ptr.get());
+                };
+            });
+        return true;
+    }
+    const std::set<nt::new_disruption::Impact*>& get_jp_impacts() const &{ return jp_impacts;}
+    std::set<nt::new_disruption::Impact*>&& get_jp_impacts() && { return std::move(jp_impacts);}
+private:
+    std::set<nt::new_disruption::Impact*> jp_impacts{};
```

Comment:
for such a simple thing, I prefer let the member be public, like that, you can move, copy or read the member freely without complicated getters.